MDNetworking/MDGameSynchronizer.cs

```diff
@@ -0,0 +1,309 @@
+    }
+
+
+    ///<summary>Calculate the player estimated offset for OS.GetTicksMSec</summary>
+    private void CalculatePlayerEstimatedTicksMSecOffset(int PeerId)
+    {
+        // TODO: Improve this to remove outliers?
+        int totalEstimate = 0;
+        InternalTicksList[PeerId].ForEach(val => totalEstimate += val);
+        totalEstimate = totalEstimate / InternalTicksList[PeerId].Count;
+        SetPlayerEstimatedOffset(PeerId, totalEstimate);
+        MDLog.Info(LOG_CAT, "Estimated OS.GetTicksMsec offset for peer [{0}] is {1} based on {2} measurements", PeerId, totalEstimate, InternalTicksList[PeerId].Count);
+    }
+
+    ///<summary>Set the estimated offset for the player</summary>
+    private void SetPlayerEstimatedOffset(int PeerId, int Estimate)
+    {
+        if (!PlayerTicksMsecOffset.ContainsKey(PeerId))
+        {
+            PlayerTicksMsecOffset.Add(PeerId, Estimate);
+        }
+        else
+        {
+            PlayerTicksMsecOffset[PeerId] = Estimate;
+        }
+    }
+
+    ///<summary>Set the ping for the player</summary>
+    private void SetPlayerPing(int PeerId, int Ping)
+    {
+        if (!PlayerPing.ContainsKey(PeerId))
+        {
+            PlayerPing.Add(PeerId, Ping);
+        }
+        else
+        {
+            PlayerPing[PeerId] = Ping;
+        }
+    }
+
+    private void OnPlayerJoinedEvent(int PeerId)
+    {
+        // Check if this is our own join message or if we are a client
+        if (PeerId == MDStatics.GetPeerId() || MDStatics.IsClient())
+        {
+            return;
+        }
+
+        // Send our first request and start timer
+        if (GetInitialMeasurementCount() > 0)
+        {
+            SendRequestToPlayer(PeerId, 1);
+        }
+    }
+
+    private void OnPlayerLeftEvent(int PeerId)
+    {
+        InternalPingList.Remove(PeerId);
+        InternalTicksList.Remove(PeerId);
+        PlayerTicksMsecOffset.Remove(PeerId);
+        PlayerPing.Remove(PeerId);
+    }
+
+    ///<summary>Starts the player ping request cycle</summary>
+    private void StartClientPingCycle(int PeerId)
+    {
+        if (IsActivePingEnabled())
+        {
+            Timer t = new Timer();
+            t.Name = "PingTimer" + PeerId;
+            t.OneShot = false;
+            t.WaitTime = GetPingInterval();
+            List<object> parameters = new List<object>();
+            parameters.Add(t);
+            parameters.Add(PeerId);
+            t.Connect("timeout", this, nameof(OnPingTimerTimeout), new Godot.Collections.Array(parameters));
+            AddChild(t);
+            t.Start();
+        }
+    }
+
+    private void OnPingTimerTimeout(Timer timer, int PeerId)
+    {
+        // Check if peer is still active
+        if (!InternalPingList.ContainsKey(PeerId) || !MDStatics.IsNetworkActive())
+        {
+            MDLog.Info(LOG_CAT, "Peer {0} has disconnected, stopping ping", PeerId);
+            timer.Stop();
+            timer.QueueFree();
```

Comment:
Should probably Remove it too, (I have a helper extension `Node.RemoveAndFree()`)